File: src/SemanticTypeForgers/forger.tests.rb
require_relative 'forger'
require 'set'

behavior = SemanticTypeForgers::AtomicBehavior.new("test")

raise "Test failed" unless behavior.getPrimitiveType(5) == 5
raise "Test failed" unless behavior.convertToPrimitive(" 123 ") == 123
raise "Test failed" unless behavior.convertToPrimitive("true") == true
raise "Test failed" unless behavior.convertToPrimitive("{}") == 0

val = behavior.proccessValue({
  "productPrice" => "10.5",
  "deliveryPrice" => 2.5
})

raise "Test failed: expected 13.0 got #{val}" unless val == 13.0

email_behavior = SemanticTypeForgers::AtomicBehavior.new("PersonEmail")
raise "Test failed" unless email_behavior.validate("test@example.com") == true
raise "Test failed" unless email_behavior.validate("invalid-email") == false

begin
  email_behavior.forge("invalid-email")
  raise "Test failed: should have thrown"
rescue => e
  raise e unless e.message.include?("Validation failed")
end

puts "Ruby tests passed"

File: src/SemanticTypeForgers/forger.tests.exs
alias SemanticTypeForgers.Forger.AtomicBehavior

behavior = AtomicBehavior.new("test")

if AtomicBehavior.get_primitive_type(5) !== 5 do
  raise "Test failed"
end

if AtomicBehavior.convert_to_primitive(" 123 ") !== 123 do
  raise "Test failed"
end

if AtomicBehavior.convert_to_primitive("true") !== true do
  raise "Test failed"
end

if AtomicBehavior.convert_to_primitive("{}") !== 0 do
  raise "Test failed"
end

val = AtomicBehavior.proccess_value(behavior, %{
  "productPrice" => "10.5",
  "deliveryPrice" => 2.5
})

if val !== 13.0 do
  raise "Test failed: expected 13.0 got #{inspect(val)}"
end

email_behavior = AtomicBehavior.new("PersonEmail")
if AtomicBehavior.validate(email_behavior, "test@example.com") !== true do
  raise "Test failed: valid email"
end

if AtomicBehavior.validate(email_behavior, "invalid-email") !== false do
  raise "Test failed: invalid email"
end

try do
  AtomicBehavior.forge(email_behavior, "invalid-email")
  raise "Test failed: should have thrown on invalid forge"
rescue
  e in RuntimeError ->
    unless String.contains?(e.message, "Validation failed") do
      raise e
    end
end

IO.puts("Elixir tests passed")
